### PR TITLE
fix(background/storage): allow extension be in multiple states

### PR DESCRIPTION
**Pull request by sidvishnoi**
## Context

Closes https://github.com/interledger/web-monetization-extension/issues/392

### CODE REVIEW

**Comment by raducristianpopa** [comment]
<h2>Extension builds preview</h2>

<table role="table">
<thead>
  <tr>
    <th>Name</th>
    <th>Link</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Latest commit</td>
    <td align="center">f24d6d998279a18d75840e5087fba088cf3649d1</td>
  </tr>
  <tr>
    <td align="center">Latest job logs</td>
    <td align="center"><a href="https://github.com/interledger/web-monetization-extension/actions/runs/9806425451">Run #9806425451</a></td>
  </tr>
  <tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/25671457882/artifacts/1670684234">Download</a></td></tr><tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/25671457882/artifacts/1670684238">Download</a></td></tr>
</tbody>
</table>

File: src/background/services/storage.ts [hunk 1]
```diff
@@ -126,23 +127,14 @@ export class StorageService {
     return false
   }
 
-  // TODO: ensure correct transitions between states, while also considering
-  // race conditions.
-  async setState(
-    state: null | Record<Exclude<Storage['state'], null>, boolean>
-  ): Promise<boolean> {
+  async setState(state: NonNullable<Storage['state']>): Promise<boolean> {
```
Comment (raducristianpopa) [inline src/background/services/storage.ts:130]: Why does the function need to return a value?

File: src/background/services/storage.ts [hunk 2]
```diff
@@ -253,9 +245,18 @@ const MIGRATIONS: Record<Storage['version'], Migration> = {
...
-      data.state = 'missing_host_permissions' satisfies Storage['state']
+      data.state = 'missing_host_permissions'
     }
 
     return [data, deleteKeys]
+  },
+  3: (data) => {
+    const state = data.state
+    if (!state) {
+      data.state = null
```
Comment (raducristianpopa) [inline src/background/services/storage.ts:256]: Can't we have it as an empty object?

File: src/shared/helpers.ts [hunk 3]
```diff
@@ -192,3 +192,12 @@ export function convert(value: bigint, source: number, target: number) {
 export function bigIntMax(a: string, b: string) {
   return BigInt(a) > BigInt(b) ? a : b
 }
+
+type Primitive = string | number | boolean | null | undefined
+
+// Warn: Not a nested object equals or a deepEquals function
```
Comment (raducristianpopa) [inline src/shared/helpers.ts:198]: 👍

File: src/background/services/storage.ts (hunk 1 above)
Comment (sidvishnoi) [reply src/background/services/storage.ts:130]: Unused for now, but can use it check if state was actually set.

File: src/background/services/storage.ts (hunk 2 above)
Comment (sidvishnoi) [reply src/background/services/storage.ts:256]: Yeah, I think that `null` optimization is unnecessary.

**Review by raducristianpopa — APPROVED** [review]

**Pull request merged**